feat(async): Add control transfer

oberien opened the pull request:
Add Transfer::control to create and submit control transfers.
In addition to the other method's parameters, it takes control setup
parameters to build the setup packet (first 8 bytes of the data buffer).
Therefore instead of `buffer: &mut [u8]` it takes `vec: &mut Vec<u8>`,
as we need to be able to prepend 8 bytes to it.

kevinmehall commented:
Thanks for starting this. I'll pick up my async branch again soon.

There's no point in taking a Vec here -- as implemented with `insert`, this will copy the entire data buffer 8 times, because it must shift over the transfer for each byte inserted.

I'm considering making the `Transfer` struct own its data buffer instead of borrowing it. That would simplify this case, because it would reserve space for the setup packet and hand out a reference to `&mut buffer[8..]` to be filled by the user.

oberien commented:
Actually there is no need to not own the buffer, as there is `Transfer::actual()` and `Transfer::buffer()`. So taking over the ownership over a `Vec<u8>` wouldn't be a problem and actually solves a lot of lifetime issues which I'm currently running into. This is that I have a struct, in which i hold the async_group, but if I want to add a transfer to it, I must make sure, that the Vec lives longer than the transfer, which is a pain in the ass :) As I need this anyways, I'm going to do that and open a new PR, which should be merged before this one, so that I can already take the Vec for this.

I'm going to optimize the 8x insert into some sort of `header.chain(vec).collect()`, which will be best case 1 copy, worst case 2-3 copies.

I also just realized, that I need to update the `actual()`-method as for control packets an offset of 8 is needed. Trying to test it atm.

oberien committed:
feat(async): Take ownership of Vec instead of slice

Instead of `&'d mut [u8]`, take ownership over `Vec<u8>`. This solves a
lot of scoping issues. Take the following example:
There is a struct `Control` with an `AsyncGroup<'a>` field, having the livetime `'a`
of `Context`. If I now want to add a new transfer to this `AsyncGroup`
in a method `Control::add_control`, I create a `Vec<u8>` in this method.
Giving it's slice to a new `Transfer`, which I then add to `AsyncGroup`,
the vec must outlive Context, so it must already be defined before I
even get the Context i.e. before I really start doing anything.

oberien commented:
- [x] I did the buffer refactor in #2 (please merge it before this one).
- [x] The header-prepender should now work with only one copy
- [x] `actual()` is updated and tested

oberien committed:
feat(async): Add control transfer

Add Transfer::control to create and submit control transfers.
In addition to the other method's parameters, it takes control setup
parameters to build the setup packet (first 8 bytes of the data buffer).
Therefore instead of `buffer: &mut [u8]` it takes `vec: &mut Vec<u8>`,
as we need to be able to prepend 8 bytes to it.

oberien committed:
fix(async): Handle OUT-Control-Packets in Transfer::actual()

Control packets with an out-direction can have a data blob. The buffer
then has 8 bytes of control-setup-header following the data blob.
`actual_length` as returned by libusb's async methods seems to be the
same as `wLength` specified in the control setup header.

Therefore if the packet is a control packet with it's direction being
`OUT`, `Transfer::actual()` checks if both lengths are the same and
returns an empty slice pointing to the end of the sent data blob.

oberien commented:
Control packets with an out-direction can have a data blob. The buffer then has 8 bytes of control-setup-header following the data blob. `actual_length` as returned by libusb's async methods seems to be the same as `wLength` specified in the control setup header.

Therefore if the packet is a control packet with it's direction being `OUT`, `Transfer::actual()` returned the data blob, which has been sent to the device. Instead an empty data blob should be returned, as the device's response didn't have a buffer. That is why I added an assertion that both lengths are the same and return an empty slice pointing to the end of the sent data blob.

I don't really know, if `(*self.transfer).actual_length` is always the same as `wLength` for control-OUT-Packets, but at least that is what I encountered with `SET_REPORT`packets. Thus I added a `TODO` to the `assert!` statement.

oberien committed:
feat(async): Add getters for Transfer::{endpoint,timeout}

oberien committed:
feat(async): Add `AsyncGroup::try_wait_any`

This function is equal to `wait_any` except it takes a Duration as
argument, which indicates, how long this functions should listen to the
device. If the timeout is reached, `None` will be returned. If during
given duration a transfer is finished, it is returned.

oberien commented:
Why did I create the 3 PRs (#1, #2, #3) in your fork 4.5 years ago instead of in the upstream? Are they still relevant? Should I move the PRs to upstream (dcuddeback/libusb-rs)?

kevinmehall commented:
dcuddeback/libusb-rs is unmaintained; you should look at https://github.com/a1ien/rusb/ instead.

My `async` branch on which this is based was never merged to either of those upstreams. It suffers from the "leakpocalypse" issue where safe code can `mem::forget` the `AsyncGroup` while the transfer is still running and violate mutable aliasing rules on the buffer. In 2020, an async abstraction should also probably be using `Future` instead of doing its own thing.

oberien commented:
Thanks for the quick response. In that case I don't think that it makes sense to keep this PR open.
